arb-gateway/src/ArbProofService.ts

```diff
@@ -0,0 +1,139 @@
+        l2RollupAddress: string,
+
+    ) {
+        this.l2Provider = l2Provider;
+        this.rollup = new Contract(
+            l2RollupAddress,
+            rollupAbi,
+            l1Provider
+        );
+        this.helper = new EVMProofHelper(l2Provider);
+    }
+
+    async getStorageAt(block: ArbProvableBlock, address: AddressLike, slot: bigint): Promise<string> {
+        return this.helper.getStorageAt(block.number, address, slot);
+    }
+
+
+    /**
+     * @dev Fetches a set of proofs for the requested state slots.
+     * @param block A `ProvableBlock` returned by `getProvableBlock`.
+     * @param address The address of the contract to fetch data from.
+     * @param slots An array of slots to fetch data for.
+     * @returns A proof of the given slots, encoded in a manner that this service's
+     *   corresponding decoding library will understand.
+     */
+    async getProofs(
+        block: ArbProvableBlock,
+        address: AddressLike,
+        slots: bigint[]
+    ): Promise<string> {
+        const proof = await this.helper.getProofs(block.number, address, slots);
+
+        return AbiCoder.defaultAbiCoder().encode(
+            [
+                'tuple(bytes32 version, bytes32 sendRoot, bytes32 blockHash,uint64 nodeIndex,bytes rlpEncodedBlock)',
+                'tuple(bytes[] stateTrieWitness, bytes[][] storageProofs)',
+            ],
+            [
+                {
+                    version:
+                        '0x0000000000000000000000000000000000000000000000000000000000000000',
+                    sendRoot: block.sendRoot,
+                    blockHash: block.blockHash,
+                    nodeIndex: block.nodeIndex,
+                    rlpEncodedBlock: block.rlpEncodedBlock
+                },
+                proof,
+            ]
+        );
+    }
+    /**
+    * Retrieves information about the latest provable block in the Arbitrum Rollup.
+    *
+    * @returns { Promise<ArbProvableBlock> } A promise that resolves to an object containing information about the provable block.
+    * @throws Throws an error if any of the underlying operations fail.
+    *
+    * @typedef { Object } ArbProvableBlock
+    * @property { string } rlpEncodedBlock - The RLP - encoded block information.
+    * @property { string } sendRoot - The send root of the provable block.
+    * @property { string } blockHash - The hash of the provable block.
+    * @property { number } nodeIndex - The index of the node corresponding to the provable block.
+    * @property { number } number - The block number of the provable block.
+    */
+    public async getProvableBlock(): Promise<ArbProvableBlock> {
+        //Retrieve the latest pending node that has been committed to the rollup.
+        const nodeIndex = await this.rollup.latestNodeCreated()
+
+        //We fetch the node created event for the node index we just retrieved.
+        const nodeEventFilter = await this.rollup.filters.NodeCreated(nodeIndex);
```

Comment:
Are any of these operations costly? Should we be caching the latest block?

Reply:
We're making two RPC calls to fetch the latest node from the L1 contract and the corresponding L2 block. Besides that, it's just event parsing. Given that the latestNodeCreated is updated roughly every hour on the mainnet, I think it could be a really good idea to cache both the Node and Block. However, there might be additional infrastructure like Redis needed to do that properly with the serverless Cloudflare worker the gateway is deployed on.

Reply:
One of the RPCs is an event filter, though, which can be slow and costly. Perhaps we can at least cache the result in-process in a variable?

Reply:
I have added a simple in-memory cache that caches the values for a particular nodeIndex. It can be easily replaced by a more sophisticated cache by implementing the IBlockCache interface